README.md
# ActiveShipping [![Build status](https://travis-ci.org/Shopify/active_shipping.svg?branch=master)](https://travis-ci.org/Shopify/active_shipping)

This library interfaces with the web services of various shipping carriers. The goal is to abstract the features that are most frequently used into a pleasant and consistent Ruby API:

- Finding shipping rates
- Registering shipments
- Tracking shipments
- Purchasing shipping labels

## Supported Shipping Carriers

* [UPS](http://www.ups.com)
* [USPS](http://www.usps.com)
* [USPS Returns](http://returns.usps.com)
* [FedEx](http://www.fedex.com)
* [Canada Post](http://www.canadapost.ca)
* [New Zealand Post](http://www.nzpost.co.nz)
* [Shipwire](http://www.shipwire.com)
* [Stamps](http://www.stamps.com)
* [Kunaki](http://www.kunaki.com)
* [Australia Post](http://auspost.com.au/)
* [Correios](http://www.correios.com.br/correios/english)


## Versions
Note: `2.x` contains breaking changes, please see the [changelog](https://github.com/Shopify/active_shipping/blob/master/CHANGELOG.md). Shopify will not be actively contributing to the `2.x` version of this gem and is looking for maintainers.

[See our releases](https://github.com/Shopify/active_shipping/releases) for past versions.

## Installation

Using bundler, add to the `Gemfile`:

```ruby
gem 'active_shipping'
```

Or stand alone:

```
$ gem install active_shipping
```


## Sample Usage

### Compare rates from different carriers

```ruby
require 'active_shipping'

# Package up a poster and a Wii for your nephew.
packages = [
  ActiveShipping::Package.new(100,               # 100 grams
                              [93,10],           # 93 cm long, 10 cm diameter
                              cylinder: true),   # cylinders have different volume calculations

  ActiveShipping::Package.new(7.5 * 16,          # 7.5 lbs, times 16 oz/lb.
                              [15, 10, 4.5],     # 15x10x4.5 inches
                              units: :imperial)  # not grams, not centimetres
 ]

 # You live in Beverly Hills, he lives in Ottawa
 origin = ActiveShipping::Location.new(country: 'US',
                                       state: 'CA',
                                       city: 'Beverly Hills',
                                       zip: '90210')

 destination = ActiveShipping::Location.new(country: 'CA',
                                            province: 'ON',
                                            city: 'Ottawa',
                                            postal_code: 'K1P 1J1')

 # Find out how much it'll be.
 ups = ActiveShipping::UPS.new(login: 'auntjudy', password: 'secret', key: 'xml-access-key')
 response = ups.find_rates(origin, destination, packages)

 ups_rates = response.rates.sort_by(&:price).collect {|rate| [rate.service_name, rate.price]}
 # => [["UPS Standard", 3936],
 #     ["UPS Worldwide Expedited", 8682],
 #     ["UPS Saver", 9348],
 #     ["UPS Express", 9702],
 #     ["UPS Worldwide Express Plus", 14502]]

 # Check out USPS for comparison...
 usps = ActiveShipping::USPS.new(login: 'developer-key')
 response = usps.find_rates(origin, destination, packages)

 usps_rates = response.rates.sort_by(&:price).collect {|rate| [rate.service_name, rate.price]}
 # => [["USPS Priority Mail International", 4110],
 #     ["USPS Express Mail International (EMS)", 5750],
 #     ["USPS Global Express Guaranteed Non-Document Non-Rectangular", 9400],
 #     ["USPS GXG Envelopes", 9400],
 #     ["USPS Global Express Guaranteed Non-Document Rectangular", 9400],
 #     ["USPS Global Express Guaranteed", 9400]]
```

Dimensions for packages are in `Height x Width x Length` order.

### Track a FedEx package

```ruby
fedex = ActiveShipping::FedEx.new(login: '999999999', password: '7777777', key: '1BXXXXXXXXXxrcB', account: '51XXXXX20')
tracking_info = fedex.find_tracking_info('tracking-number', carrier_code: 'fedex_ground') # Ground package

tracking_info.shipment_events.each do |event|
  puts "#{event.name} at #{event.location.city}, #{event.location.state} on #{event.time}. #{event.message}"
end
# => Package information transmitted to FedEx at NASHVILLE LOCAL, TN on Thu Oct 23 00:00:00 UTC 2008.
# Picked up by FedEx at NASHVILLE LOCAL, TN on Thu Oct 23 17:30:00 UTC 2008.
# Scanned at FedEx sort facility at NASHVILLE, TN on Thu Oct 23 18:50:00 UTC 2008.
# Departed FedEx sort facility at NASHVILLE, TN on Thu Oct 23 22:33:00 UTC 2008.
# Arrived at FedEx sort facility at KNOXVILLE, TN on Fri Oct 24 02:45:00 UTC 2008.
# Scanned at FedEx sort facility at KNOXVILLE, TN on Fri Oct 24 05:56:00 UTC 2008.
# Delivered at Knoxville, TN on Fri Oct 24 16:45:00 UTC 2008. Signed for by: T.BAKER
```

## Carrier specific notes

### FedEx connection

The `:login` key passed to `ActiveShipping::FedEx.new()` is really the FedEx meter number, not the FedEx login.

When developing with test credentials, be sure to pass `test: true` to `ActiveShipping::FedEx.new()`.


## Tests

You can run the unit tests with:

```
bundle exec rake test:unit
```

and the remote tests with:

```
bundle exec rake test:remote
```

The unit tests mock out requests and responses so that everything runs locally, while the remote tests actually hit the carrier servers. For the remote tests, you'll need valid test credentials for any carriers' tests you want to run. The credentials should go in [`~/.active_shipping/credentials.yml`](https://github.com/Shopify/active_shipping/blob/master/test/credentials.yml). For some carriers, we have public credentials you can use for testing in `.travis.yml`. Remote tests missing credentials will be skipped.


## Contributing

See [CONTRIBUTING.md](https://github.com/Shopify/active_shipping/blob/master/CONTRIBUTING.md).

We love getting pull requests! Anything from new features to documentation clean up.

If you're building a new carrier, a good place to start is in the [`Carrier` base class](https://github.com/Shopify/active_shipping/blob/master/lib/active_shipping/carrier.rb).

It would also be good to familiarize yourself with [`Location`](https://github.com/Shopify/active_shipping/blob/master/lib/active_shipping/location.rb), [`Package`](https://github.com/Shopify/active_shipping/blob/master/lib/active_shipping/package.rb), and [`Response`](https://github.com/Shopify/active_shipping/blob/master/lib/active_shipping/response.rb).

You can use the [`test/console.rb`](https://github.com/Shopify/active_shipping/blob/master/test/console.rb) to do some local testing against real endpoints.

To log requests and responses, just set the `logger` on your Carrier class to some kind of `Logger` object:

```ruby
ActiveShipping::USPS.logger = Logger.new(STDOUT)
```

### Anatomy of a pull request

Any new features or carriers should have passing unit _and_ remote tests. Look at some existing carriers as examples.

When opening a pull request, include description of the feature, why it exists, and any supporting documentation to explain interaction with carriers.


### How to contribute

1. Fork it ( https://github.com/Shopify/active_shipping/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# ActiveShipping CHANGELOG

=======
### v2.2.1
- Add support for Rails 5.2

### v2.1.1
- Update README to clarify Shopify's involvement moving forward with v2.x

### v2.1.0
- Add email field to Location

### v2.0.0
- Drop support for < ruby 2.2, support ruby 2.4
- Drop support for < Rails 4.2.
- BREAKING CHANGE: Use shopify/measured instead of shopify/quantified for dimensions and units.

### v1.14.2
- Make saturday delivery an option for fedex

### v1.14.1
- Fix parsing of CanadaPostPWS service options response.

### v1.14.0
- Update Correios default services list.
- Fix CanadaPostPWS from generating an empty options tag.
- Allow contract-number on CanadaPost merchant detail's response to be nil.
- Fix a flakey UPS remote test that would fail only on Fridays.

### v1.13.4
- Upcase postal code for CanadaPostPWS
- Fix failing USPS test
- add .byebug_history to .gitignore

### v1.13.3
- CanadaPostPWS no longer modifies locations passed to it

### v1.13.2
- Bump active_utils to 3.3.1
- Allow activesupport <5.2.0

### v1.13.1
- Fix up UPS tracker parsing for Kosovo (KV)

### v1.13.0
- Add default location for CanadaPost PWS
- Patch UPS to use old Kosovo country code
- Add option to not include tax in rates for CP PWS

### v1.12.1
- Explicitly set ssl_version for USPS
- Strip 9 digit origin zip code for USPS world rate requests

### v1.12.0
- Update active_utils dependency to v3.3.0

### v1.8.6
- Fix UPS TrackResponse with no status code
- Stop FedEx from raising for successful responses with no statuses
- Raise appropriate exception response for FedEx errors

### v1.8.5
- Fix UPS TrackResponse parsing for missing elements

### v1.8.4
- Add price details to rate estimates
- Fix encoding for UPS responses

### v1.8.3
- Add description field to rate estimates

### v1.8.2
- Add option for FedEx label format
- Fix kunaki remote tests broken due to more shipping options

### v1.6.1
- Fix FedEx ShipmentEvents to include event type
- Skip broken Canada Post remote tests

### v1.6.0
- Update active_utils dependency to v3.2.0

### v1.5.0
- Fix Kunaki remote test
- Fix credentials for Canada Post PWS remote tests
- Add compare price to rate estimate options
- Add phone required to rate estimate options
- Update active_utils dependency to v3.1.0

### v1.4.3

- Fix UPS SurePost for < 1 pound packages
- Use status type code in UPS tracking
- Fix USPS and Fedex remote tests

### v1.4.2

- Fix USPS rates for commercial shipments

### v1.4.1

- Raise error on invalid status code with FedEx
- Fix USPS tracking to certain countries
- Fix USPS tracking of events with no time
- Fix USPS batch tracking error messages.
- Fix FedEx logging exception

### v1.4.0

- Added support for USPS merchant returns service
- Added support for UPS SurePost
- Added support for UPS third party billing
- Fix FedEx tracking response errors
- Add rake console command for development

### v1.3.0

- Support voiding labels on UPS
- Parse FedEx ground delivery dates
- Add maximum address length field
- Fix UPS unknown country code when using SUREPOST

### v1.2.2

- Fix "RECTANGULAR" errors with small USPS US->US package rate requests
- Fix error tracking USPS packages to other countries
- Fix USPS rate requests to destinations with only a country.

### v1.2.1

- Fix compatibility with latest USPS WEBTOOLS international rate schema changes.

### v1.2.0

- Added support for buying labels with FedEx
- Added support for batched tracking requests with USPS

### v1.1.3

- Handle ZIP+4 numbers in USPS tracking
- Add a field for rate estimate references
- Add support for UPS mail innovations tracking option

### v1.1.2

- Fix finding of error descriptions for USPS tracking

### v1.1.1

- Fix bug with USPS tracking not handling optional fields being absent.

### v1.1.0

- USPS: Allows package tracking disambiguation and exposes predicted arrival date and event codes.

### v1.0.0

- **BREAKING CHANGE:** Change namespace from `ActiveMerchant::Shipping` to `ActiveShipping`
- Drop support for Ruby 1.8 and Ruby 1.9.
- Drop support for ActiveSupport < 3.2, support up to ActiveSupport 4.2.
- Updated Fedex to use latest API version for tracking
- Various improvements to UPS carrier implementation.
- Small bugfixes in USPS carrier implementation.
- Various small bugfixes in XML handling for several carriers.
- Rewite all carriers to use nokogiri for XML parsing and generating.
- Bump `active_utils` dependency to require 3.x to avoid conflicts with `ActiveMerchant`.
- Extracted `quantified` into separate gem.
- Removed vendored `XmlNode` library.
- Removed `builder` dependency.
- Improved test setup that allows running functional tests on CI.
- Improved documentation of the abstraction API.

### v0.10.1

 - Canada Post PWS: Makes wrapper act more consistently with the rest of the API [jnormore]
 - UPS: Adds insurance charge to package object declarations [pbonnell]
 - USPS: Improves how unavailable delivery information is handled [cyu]
 - Shipment Packer: Prevents packing errors and consistently return an array when packing [christianblais]
 - General: Improves tests such that they work with ruby 2.0 [Sirupsen]

### 2011/04/21

* USPS updated to use new APIs [james]
* new :gift boolean option for Package [james]
* Location's :address_type can be "po_box" [james]

### Earlier

* New Zealand Post [AbleTech]
* Include address name for rate requests to Shipwire if provided [dennis]
* Add support for address name to Location [dennis]
* Add fix for updated USPS API to strip any encoded html and trailing asterisks from rate names [dennis]
* Add carrier CanadaPost [william]
* Update FedEx rates and added ability to auto-generate rate name from code that gets returned by FedEx [dennis]

lib/active_shipping/shipping/carriers/usps.rb
require 'cgi'

module ActiveMerchant
  module Shipping
    
    # After getting an API login from USPS (looks like '123YOURNAME456'),
    # run the following test:
    # 
    # usps = USPS.new(:login => '123YOURNAME456', :test => true)
    # usps.valid_credentials?
    #
    # This will send a test request to the USPS test servers, which they ask you
    # to do before they put your API key in production mode.
    class USPS < Carrier
      self.retry_safe = true
      
      cattr_reader :name
      @@name = "USPS"
      
      LIVE_DOMAIN = 'production.shippingapis.com'
      LIVE_RESOURCE = '/ShippingAPI.dll'
      
      TEST_DOMAINS = { #indexed by security; e.g. TEST_DOMAINS[USE_SSL[:rates]]
        true => 'secure.shippingapis.com',
        false => 'testing.shippingapis.com'
      }
      
      TEST_RESOURCE = '/ShippingAPITest.dll'
      
      API_CODES = {
        :us_rates => 'RateV3',
        :world_rates => 'IntlRate',
        :test => 'CarrierPickupAvailability'
      }
      USE_SSL = {
        :us_rates => false,
        :world_rates => false,
        :test => true
      }
      CONTAINERS = {
        :envelope => 'Flat Rate Envelope',
        :box => 'Flat Rate Box'
      }
      MAIL_TYPES = {
        :package => 'Package',
        :postcard => 'Postcards or aerogrammes',
        :matter_for_the_blind => 'Matter for the blind',
        :envelope => 'Envelope'
      }
      PACKAGE_PROPERTIES = {
        'ZipOrigination' => :origin_zip,
        'ZipDestination' => :destination_zip,
        'Pounds' => :pounds,
        'Ounces' => :ounces,
        'Container' => :container,
        'Size' => :size,
        'Machinable' => :machinable,
        'Zone' => :zone,
        'Postage' => :postage,
        'Restrictions' => :restrictions
      }
      POSTAGE_PROPERTIES = {
        'MailService' => :service,
        'Rate' => :rate
      }
      US_SERVICES = {
        :first_class => 'FIRST CLASS',
        :priority => 'PRIORITY',
        :express => 'EXPRESS',
        :bpm => 'BPM',
        :parcel => 'PARCEL',
        :media => 'MEDIA',
        :library => 'LIBRARY',
        :all => 'ALL'
      }
      
      # TODO: get rates for "U.S. possessions and Trust Territories" like Guam, etc. via domestic rates API: http://www.usps.com/ncsc/lookups/abbr_state.txt
      # TODO: figure out how USPS likes to say "Ivory Coast"
      #
      # Country names:
      # http://pe.usps.gov/text/Imm/immctry.htm
      COUNTRY_NAME_CONVERSIONS = {
        "BA" => "Bosnia-Herzegovina",
        "CD" => "Congo, Democratic Republic of the",
        "CG" => "Congo (Brazzaville),Republic of the",
        "CI" => "Côte d'Ivoire (Ivory Coast)",
        "CK" => "Cook Islands (New Zealand)",
        "FK" => "Falkland Islands",
        "GB" => "Great Britain and Northern Ireland",
        "GE" => "Georgia, Republic of",
        "IR" => "Iran",
        "KN" => "Saint Kitts (St. Christopher and Nevis)",
        "KP" => "North Korea (Korea, Democratic People's Republic of)",
        "KR" => "South Korea (Korea, Republic of)",
        "LA" => "Laos",
        "LY" => "Libya",
        "MC" => "Monaco (France)",
        "MD" => "Moldova",
        "MK" => "Macedonia, Republic of",
        "MM" => "Burma",
        "PN" => "Pitcairn Island",
        "RU" => "Russia",
        "SK" => "Slovak Republic",
        "TK" => "Tokelau (Union) Group (Western Samoa)",
        "TW" => "Taiwan",
        "TZ" => "Tanzania",
        "VA" => "Vatican City",
        "VG" => "British Virgin Islands",
        "VN" => "Vietnam",
        "WF" => "Wallis and Futuna Islands",
        "WS" => "Western Samoa"
      }

      def self.size_code_for(package)
        total = package.inches(:length) + package.inches(:girth)
        if total <= 84
          return 'REGULAR'
        elsif total <= 108
          return 'LARGE'
        else # <= 130
          return 'OVERSIZE'
        end
      end
      
      # from info at http://www.usps.com/businessmail101/mailcharacteristics/parcels.htm
      # 
      # package.options[:books] -- 25 lb. limit instead of 35 for books or other printed matter.
      #                             Defaults to false.
      def self.package_machinable?(package, options={})
        at_least_minimum =  package.inches(:length) >= 6.0 &&
                            package.inches(:width) >= 3.0 &&
                            package.inches(:height) >= 0.25 &&
                            package.ounces >= 6.0
        at_most_maximum  =  package.inches(:length) <= 34.0 &&
                            package.inches(:width) <= 17.0 &&
                            package.inches(:height) <= 17.0 &&
                            package.pounds <= (package.options[:books] ? 25.0 : 35.0)
        at_least_minimum && at_most_maximum
      end
      
      def requirements
        [:login]
      end
      
      def find_rates(origin, destination, packages, options = {})
        options = @options.merge(options)
        
        origin = Location.from(origin)
        destination = Location.from(destination)
        packages = Array(packages)
        
        #raise ArgumentError.new("USPS packages must originate in the U.S.") unless ['US',nil].include?(origin.country_code(:alpha2))
        
        
        # domestic or international?
        
        response = if ['US',nil].include?(destination.country_code(:alpha2))
          us_rates(origin, destination, packages, options)
        else
          world_rates(origin, destination, packages, options)
        end
      end
      
      def valid_credentials?
        # Cannot test with find_rates because USPS doesn't allow that in test mode
        test_mode? ? canned_address_verification_works? : super
      end
      
      def maximum_weight
        Mass.new(70, :pounds)
      end
      
      protected
      
      def us_rates(origin, destination, packages, options={})
        request = build_us_rate_request(packages, origin.zip, destination.zip, options)
         # never use test mode; rate requests just won't work on test servers
        parse_rate_response origin, destination, packages, commit(:us_rates,request,false), options
      end
      
      def world_rates(origin, destination, packages, options={})
        request = build_world_rate_request(packages, destination.country)
         # never use test mode; rate requests just won't work on test servers
        parse_rate_response origin, destination, packages, commit(:world_rates,request,false), options
      end
      
      # Once the address verification API is implemented, remove this and have valid_credentials? build the request using that instead.
      def canned_address_verification_works?
        request = "%3CCarrierPickupAvailabilityRequest%20USERID=%22#{@options[:login]}%22%3E%20%0A%3CFirmName%3EABC%20Corp.%3C/FirmName%3E%20%0A%3CSuiteOrApt%3ESuite%20777%3C/SuiteOrApt%3E%20%0A%3CAddress2%3E1390%20Market%20Street%3C/Address2%3E%20%0A%3CUrbanization%3E%3C/Urbanization%3E%20%0A%3CCity%3EHouston%3C/City%3E%20%0A%3CState%3ETX%3C/State%3E%20%0A%3CZIP5%3E77058%3C/ZIP5%3E%20%0A%3CZIP4%3E1234%3C/ZIP4%3E%20%0A%3C/CarrierPickupAvailabilityRequest%3E%0A"
        # expected_hash = {"CarrierPickupAvailabilityResponse"=>{"City"=>"HOUSTON", "Address2"=>"1390 Market Street", "FirmName"=>"ABC Corp.", "State"=>"TX", "Date"=>"3/1/2004", "DayOfWeek"=>"Monday", "Urbanization"=>nil, "ZIP4"=>"1234", "ZIP5"=>"77058", "CarrierRoute"=>"C", "SuiteOrApt"=>"Suite 777"}}
        xml = REXML::Document.new(commit(:test, request, true))
        xml.get_text('/CarrierPickupAvailabilityResponse/City').to_s == 'HOUSTON' &&
        xml.get_text('/CarrierPickupAvailabilityResponse/Address2').to_s == '1390 Market Street'
      end
      
      # options[:service] --    One of [:first_class, :priority, :express, :bpm, :parcel,
      #                          :media, :library, :all]. defaults to :all.
      # options[:container] --  One of [:envelope, :box]. defaults to neither (this field has
      #                          special meaning in the USPS API).
      # options[:books] --      Either true or false. Packages of books or other printed matter
      #                          have a lower weight limit to be considered machinable.
      # package.options[:machinable] -- Either true or false. Overrides the detection of
      #                                  "machinability" entirely.
      def build_us_rate_request(packages, origin_zip, destination_zip, options={})
        packages = Array(packages)
        request = XmlNode.new('RateV3Request', :USERID => @options[:login]) do |rate_request|
          packages.each_with_index do |p,id|
            rate_request << XmlNode.new('Package', :ID => id.to_s) do |package|
              package << XmlNode.new('Service', US_SERVICES[options[:service] || :all])
              package << XmlNode.new('ZipOrigination', strip_zip(origin_zip))
              package << XmlNode.new('ZipDestination', strip_zip(destination_zip))
              package << XmlNode.new('Pounds', 0)
              package << XmlNode.new('Ounces', "%0.1f" % [p.ounces,1].max)
              if p.options[:container] and [nil,:all,:express,:priority].include? p.service
                package << XmlNode.new('Container', CONTAINERS[p.options[:container]])
              end
              package << XmlNode.new('Size', USPS.size_code_for(p))
              package << XmlNode.new('Width', p.inches(:width))
              package << XmlNode.new('Length', p.inches(:length))
              package << XmlNode.new('Height', p.inches(:height))
              package << XmlNode.new('Girth', p.inches(:girth))
              is_machinable = if p.options.has_key?(:machinable)
                p.options[:machinable] ? true : false
              else
                USPS.package_machinable?(p)
              end
              package << XmlNode.new('Machinable', is_machinable.to_s.upcase)
            end
          end
        end
        URI.encode(save_request(request.to_s))
      end
      
      # important difference with international rate requests:
      # * services are not given in the request
      # * package sizes are not given in the request
      # * services are returned in the response along with restrictions of size
      # * the size restrictions are returned AS AN ENGLISH SENTENCE (!?)
      #
      # 
      # package.options[:mail_type] -- one of [:package, :postcard, :matter_for_the_blind, :envelope].
      #                                 Defaults to :package.
      def build_world_rate_request(packages, destination_country)
        country = COUNTRY_NAME_CONVERSIONS[destination_country.code(:alpha2).first.value] || destination_country.name
        request = XmlNode.new('IntlRateRequest', :USERID => @options[:login]) do |rate_request|
          packages.each_index do |id|
            p = packages[id]
            rate_request << XmlNode.new('Package', :ID => id.to_s) do |package|
              package << XmlNode.new('Pounds', 0)
              package << XmlNode.new('Ounces', [p.ounces,1].max.ceil) #takes an integer for some reason, must be rounded UP
              package << XmlNode.new('MailType', MAIL_TYPES[p.options[:mail_type]] || 'Package')
              package << XmlNode.new('ValueOfContents', p.value / 100.0) if p.value && p.currency == 'USD'
              package << XmlNode.new('Country') do |node|
                node.cdata = country
              end
            end
          end
        end
        URI.encode(save_request(request.to_s))
      end
      
      def parse_rate_response(origin, destination, packages, response, options={})
        success = true
        message = ''
        rate_hash = {}
        
        xml = REXML::Document.new(response)
        
        if error = xml.elements['/Error']
          success = false
          message = error.elements['Description'].text
        else
          xml.elements.each('/*/Package') do |package|
            if package.elements['Error']
              success = false
              message = package.get_text('Error/Description').to_s
              break
            end
          end
          
          if success
            rate_hash = rates_from_response_node(xml, packages)
            unless rate_hash
              success = false
              message = "Unknown root node in XML response: '#{root_node_name}'"
            end
          end
          
        end
        
        rate_estimates = rate_hash.keys.map do |service_name|
          RateEstimate.new(origin,destination,@@name,"USPS #{service_name}",
                                    :package_rates => rate_hash[service_name][:package_rates],
                                    :service_code => rate_hash[service_name][:service_code],
                                    :currency => 'USD')
        end
        rate_estimates.reject! {|e| e.package_count != packages.length}
        rate_estimates = rate_estimates.sort_by(&:total_price)
        
        RateResponse.new(success, message, ActiveMerchant.parse_xml(response), :rates => rate_estimates, :xml => response, :request => last_request, :log_xml => options[:log_xml])
      end
      
      def rates_from_response_node(response_node, packages)
        rate_hash = {}
        return false unless (root_node = response_node.elements['/IntlRateResponse | /RateV3Response'])
        domestic = (root_node.name == 'RateV3Response')
        
        domestic_elements = ['Postage', 'CLASSID', 'MailService', 'Rate']
        international_elements = ['Service', 'ID', 'SvcDescription', 'Postage']
        service_node, service_code_node, service_name_node, rate_node = domestic ? domestic_elements : international_elements
        
        root_node.each_element('Package') do |package_node|
          package_index = package_node.attributes['ID'].to_i
          
          package_node.each_element(service_node) do |service_response_node|
            service_name = service_response_node.get_text(service_name_node).to_s
            
            # aggregate specific package rates into a service-centric RateEstimate
            # first package with a given service name will initialize these;
            # later packages with same service will add to them
            this_service = rate_hash[service_name] ||= {}
            this_service[:service_code] ||= service_response_node.attributes[service_code_node]
            package_rates = this_service[:package_rates] ||= []
            this_package_rate = {:package => (this_package = packages[package_index]),
                                 :rate => Package.cents_from(service_response_node.get_text(rate_node).to_s.to_f)}
            
            package_rates << this_package_rate if package_valid_for_service(this_package,service_response_node)
          end
        end
        rate_hash
      end
      
      def package_valid_for_service(package, service_node)
        return true if service_node.elements['MaxWeight'].nil?
        max_weight = service_node.get_text('MaxWeight').to_s.to_f
        name = service_node.get_text('SvcDescription | MailService').to_s.downcase
        
        if name =~ /flat.rate.box/ #domestic or international flat rate box
          # flat rate dimensions from http://www.usps.com/shipping/flatrate.htm
          return (package_valid_for_max_dimensions(package,
                      :weight => max_weight, #domestic apparently has no weight restriction
                      :length => 11.0,
                      :width => 8.5,
                      :height => 5.5) or
                 package_valid_for_max_dimensions(package,
                      :weight => max_weight,
                      :length => 13.625,
                      :width => 11.875,
                      :height => 3.375))
        elsif name =~ /flat.rate.envelope/
          return package_valid_for_max_dimensions(package,
                      :weight => max_weight,
                      :length => 12.5,
                      :width => 9.5,
                      :height => 0.75)
        elsif service_node.elements['MailService'] # domestic non-flat rates
          return true
        else #international non-flat rates
          # Some sample english that this is required to parse:
          #
          # 'Max. length 46", width 35", height 46" and max. length plus girth 108"'
          # 'Max. length 24", Max. length, height, depth combined 36"'
          # 
          sentence = CGI.unescapeHTML(service_node.get_text('MaxDimensions').to_s)
          tokens = sentence.downcase.split(/[^\d]*"/).reject {|t| t.empty?}
          max_dimensions = {:weight => max_weight}
          single_axis_values = []
          tokens.each do |token|
            axis_sum = [/length/,/width/,/height/,/depth/].sum {|regex| (token =~ regex) ? 1 : 0}
            unless axis_sum == 0
              value = token[/\d+$/].to_f 
              if axis_sum == 3
                max_dimensions[:length_plus_width_plus_height] = value
              elsif token =~ /girth/ and axis_sum == 1
                max_dimensions[:length_plus_girth] = value
              else
                single_axis_values << value
              end
            end
          end
          single_axis_values.sort!.reverse!
          [:length, :width, :height].each_with_index do |axis,i|
            max_dimensions[axis] = single_axis_values[i] if single_axis_values[i]
          end
          return package_valid_for_max_dimensions(package, max_dimensions)
        end
      end
      
      def package_valid_for_max_dimensions(package,dimensions)
        valid = ((not ([:length,:width,:height].map {|dim| dimensions[dim].nil? || dimensions[dim].to_f >= package.inches(dim).to_f}.include?(false))) and
                (dimensions[:weight].nil? || dimensions[:weight] >= package.pounds) and
                (dimensions[:length_plus_girth].nil? or
                    dimensions[:length_plus_girth].to_f >=
                    package.inches(:length) + package.inches(:girth)) and
                (dimensions[:length_plus_width_plus_height].nil? or
                    dimensions[:length_plus_width_plus_height].to_f >=
                    package.inches(:length) + package.inches(:width) + package.inches(:height)))

        return valid
      end
      
      def commit(action, request, test = false)
        ssl_get(request_url(action, request, test))
      end
      
      def request_url(action, request, test)
        scheme = USE_SSL[action] ? 'https://' : 'http://'
        host = test ? TEST_DOMAINS[USE_SSL[action]] : LIVE_DOMAIN
        resource = test ? TEST_RESOURCE : LIVE_RESOURCE
        "#{scheme}#{host}/#{resource}?API=#{API_CODES[action]}&XML=#{request}"
      end
      
      def strip_zip(zip)
        zip.to_s.scan(/\d{5}/).first || zip
      end
      
    end
  end
end
